elastic/elastic-agent
# [Elastic Agent]-PR- Introduce log message for not supported annotations for Hints based autodiscover

<del>Note: Dont merge  until https://github.com/elastic/elastic-agent-autodiscover/pull/81 is merged.</del>
<del>We need to udpate go.mod of the elastic-agent with v0.6.9</del> PR is merged

- Enhancement

## What does this PR do?

Relates to:
- https://github.com/elastic/elastic-agent/issues/3063
- https://github.com/elastic/elastic-agent/issues/3064
- https://github.com/elastic/elastic-agent-autodiscover/pull/81

## Why is it important?

It adds a helpful feature for those that use Hints Autodiscovery. It can be helpful to identify configuration issues

- [ ] I have added tests that prove my fix is effective or that my feature works (Some tests updated [here](https://github.com/elastic/elastic-agent-autodiscover/pull/81/files#diff-2f7d9a00d75cb68d90813219822383fdda49c60b39cc05f7d2713f3d9e18a84a))


## How to test this PR locally

Follow same instructions as https://github.com/elastic/elastic-agent/pull/3107
1. Checkout this branch
2. Update hints.go file of your <home>/go/pkg/mod/github.com/elastic/elastic-agent-autodiscover@v0.6.8/utils as per [hints.go](https://github.com/elastic/elastic-agent-autodiscover/pull/81/files#diff-99d5db949c88160c717cd1240bed90ec4da0d96b04cd287de3ca53d0cd90e4bf)
3. Build elastic agent image
```bash
EXTERNAL=true DEV=true SNAPSHOT=true PLATFORMS=linux/arm64 PACKAGES=docker mage package
cd build/package/elastic-agent/elastic-agent-linux-arm64.docker/docker-build
docker build -t custom-agent-image .\n
kind load docker-image custom-agent-image:latest\n
```
4. Install elastic-agent-standalone with hints enabled. Sample here
[elastic-agent-standalone.txt](https://github.com/elastic/elastic-agent/files/14509781/elastic-agent-standalone.txt)

5. Configure some "wrong" co.elastic.hints annotation inside a pod. 
```yaml
annotations:        
        co.elastic.hints/enabled: 'true'
        co.elastic.hints/package: "container_logs"
        co.elastic.hints/hostssssss: '128.30.40.50'
```

7. Warning messages should be available 

## Related issues

- Relates https://github.com/elastic/elastic-agent/issues/3064

## Logs

Deploy following pod with wrong annotations hosts and steam:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: nginx
  name: nginx
  namespace: default
spec:
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
      annotations:
        co.elastic.hints/enabled: 'true'
        co.elastic.hints/hosts: '128.30.40.50'
        co.elastic.hints/steam: 'stdout'
    spec:
      containers:
      - image: nginx
        name: nginx
```

```bash
{"log.level":"warn","@timestamp":"2024-03-07T13:15:22.599Z","log.logger":"composable.providers.kubernetes","log.origin":{"file.name":"kubernetes/pod.go","file.line":219},"message":"provided hint: co.elastic/hints/hosts is not in the supported list for pod nginx-68d79864cb-rgjbb in namespace default","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2024-03-07T13:15:22.599Z","log.logger":"composable.providers.kubernetes","log.origin":{"file.name":"kubernetes/pod.go","file.line":219},"message":"provided hint: co.elastic/hints/steam is not in the supported list for pod nginx-68d79864cb-rgjbb in namespace default","log":{"source":"elastic-agent"},"ecs.version":"1.6.0"}
```

Check that only annotations that start with co.elastic are the ones that are checked for correctness.

## Screenshots

![Screenshot 2024-03-07 at 3 21 19 PM](https://github.com/elastic/elastic-agent/assets/1685415/2ebce491-e523-4b8e-bb53-d730310f3b23)
